hazelcast-jet-core/src/main/java/com/hazelcast/jet/pipeline/Pipeline.java

```diff
@@ -99,9 +100,24 @@ static Pipeline create() {
     @Nonnull
     DAG toDag();
 
+    /**
+     * Transforms the pipeline into a Jet DAG, which can be submitted for
+     * execution to a Jet instance.
+     */
+    @Nonnull
+    DAG toDag(Context context);
```

Comment:
1. Do we need this as a public API? 
2. We miss the `@since` tag here
3. The name `Context` suggests that there will be more methods in the future, but those methods will have to be `default`. Context object instances are normally provided by the environment, not by the user. Why don't we just use `int defaultLocalParallelism` argument?
4. Instead of a new method we could maybe deprecate the `toDag()` method. It's not a supported way to use Jet.
5. If we don't provide this method, I think we can also remove the `AbstractTransform.determineLocalParallelism`, which is almost a clone of `Vertex.determineLocalParallelism`. We can keep just -1 for LP and the vertex will take care. But in this point maybe I missed something, I'm not sure.

Reply:
Yes, this should be retracted from the public API. Then it will be easy to decide not to take a config object. 

Having default implementations isn't necessary, I think, since Jet would be the only implementor.

Reply:
5) For now, this LP determination process looks like the clone of the Vertex inside, but it will later be used to propagate the LP of a transform to its downstream and will be modified for it. The determineLocalParallelism() in Vertex did not allow such a change because it could not see LP of the upstream vertices. After adding this method to Transforms, I also tried to remove the determineLocalParallelism in Vertex, but the problem arises when the client directly submits a DAG containing -1 LP vertex as a job definition.
For the other items, I will apply the fix specified by Marko.

Reply:
>it will later be used to propagate the LP of a transform to its downstream 

We don't need it now, will it be needed for some future feature?

Reply:
yes, we don't need this for this relocation but it will be needed for the future feature.